source/develop/release-management/features/storage/4k-blocks.md

```diff
@@ -0,0 +1,93 @@
+
+*   Target Release: 4.3
+*   Status: Planning
+*   Last updated: July 31, 2018
+
+# Implementation details
+
+VDSM heavily relies on 512 block support and it is hardcoded to a lots of places.
+
+The general approach is to refactor all block size dependent code to use block size from the storage domain, while keeping it hardcoded to 512b in it and in a second step, implement block size detection at the SD level and replace hardcoded value with actual value.
+
+Subsystem specific notes are below:
+
+## Sanlock
+
+Sanlock is used by oVirt for locking and fencing shared resources. Latest released sanlock have native support for 4k devices, but, unfortunately, at the moment it work only with block devices, not with filesystems. Speaking precisely, sanlock on filesystems always assumes, that block size of the device is 512b. With patch [1] it will be possible, to specify desired block size, but it is still not able to detect block size of the filesystem backing device.
+
+## Blocksize detection
+
+VDSM have to detect block size during storage domain creation. For block device backed storage domains sanlock or blkid could be used. For filesystem backed storage domains VDSM can actually try to make 512b atomic operation and switch to 4k only if it fails. Something like dd if=/fs/under/test/file of=/dev/null bs=512 count=1 iflag=direct
+
+## Metadata
+
+VDSM needs to know size of block of the Storage domain and the best way to store it is metadata. Metadata for block device backed domains already includes corresponding fields:
+
+* DMDK_LOGBLKSIZE
+* DMDK_PHYBLKSIZE
+
+But metadata for the filesystem backed storage domains do not have those fields.  Block/FS domains also differ in two other fields:
+
+* DMDK_VGUUID - at block backed SD
+* REMOTE_PATH - at filesystem backed SD
+
+As those fields are unused, new version of storage domains needs to implemented, removing unused fields from the metadata and adding block size fields to the filesystems based SD, thus unifying metadata.
+
+This also means, that manifests for both types of domains should include block size field. Block based domains already do that, but filesystem domains do not
+
+## Locking
+
+xlease modules uses BLOCK_SIZE constant and replacing it with externally supplied block size should make it 4k friendly.
+
+SDM and domain locking use hardcoded block size and it should be fixed.
+
+## Volumes
+
+VolumeManifest have support for variable block sizes, but implementations hard code it to 512b blocks. SD volume creation code should set block size at the Volume.
+
+Volume creation flow is also handled by sdm.VolumeArtifacts, that uses BLOCK_SIZE constants, which should be replaced with block size from the volume manifest. 
+
+BlockVolume and FileVolume implementations must use block size from the VolumeManifest. That should be done in two steps - first move implementations to use hardcoded block size at VolumeManifest and after that make block size adjustable at the VolumeManifest side.
```

Comment:
The other way around, volume code should use block size from the domain. We should probably move volume creation code to the storage domain, where it belongs.